File: api/src/models/UserModel.py
from datetime import date
import uuid
from typing import List, Optional, TYPE_CHECKING
from src.database import Base
from sqlalchemy import Boolean, CheckConstraint, ForeignKey, Integer, String, Date, UUID
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import Mapped, mapped_column, relationship
from src.models.SurveyUserModel import survey_user


if TYPE_CHECKING:
    from src.models.UserFcmTokenModel import UserFcmToken
    from src.models.SurveyModel import Survey
    from src.models.OrganizationModel import Organization
    from src.models.AnswerModel import Answer





class User(Base):
    __tablename__ = "users"

    id: Mapped[uuid.UUID] = mapped_column(UUID, primary_key=True, index=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(50), nullable=True)
    lastname: Mapped[str] = mapped_column(String(100), nullable=True)
    email: Mapped[str] = mapped_column(String(100), nullable=False, unique=True)
    password: Mapped[str] = mapped_column(String(100), nullable=False)
    role: Mapped[str] = mapped_column(String(50), nullable=False)
    birthdate: Mapped[date] = mapped_column(Date, nullable=True)
    gender: Mapped[str] = mapped_column(String(50), nullable=True)
    allow_notifications: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)


    organization_id: Mapped[uuid.UUID] = mapped_column(UUID, ForeignKey("organizations.id"), nullable=True)
    
    

    organization: Mapped["Organization"] = relationship(back_populates="users", lazy="selectin")
    surveys_assigned: Mapped[Optional[List["Survey"]]]=relationship(secondary=survey_user, back_populates="assigned_users", lazy="selectin")
    surveys_owned: Mapped[Optional[List["Survey"]]] = relationship(back_populates="owner", cascade="all, delete", lazy="selectin")
    answers: Mapped[Optional[List["Answer"]]] = relationship(back_populates="participant", cascade="all, delete", lazy="selectin")
    fcm_tokens: Mapped[Optional[List["UserFcmToken"]]] = relationship(back_populates="user", cascade="all, delete", lazy="selectin")

    __table_args__ = (
        CheckConstraint("role IN ('researcher', 'participant', 'admin')", name="role_check"),
        CheckConstraint("gender IN ('male', 'female', 'other')", name="gender_check"),
        CheckConstraint("birthDate <=CURRENT_DATE", name="birthDate_check"),
    )

    @hybrid_property
    def age(self):
        today = date.today()
        if self.birthdate: 
            return today.year - self.birthdate.year - ((today.month, today.year) < (self.birthdate.month, self.birthdate.year))
        return None
    

File: client/surbased/lib/src/user/domain/user_model.dart
import 'dart:convert';

User userFromJson(String str) => User.fromJson(json.decode(str));

String userToJson(User data) => json.encode(data.toJson());

class User {
  final String email;
  final String? name;
  final String? lastname;
  final String role;
  final String id;
  final String? organizationId;
  final DateTime? birthdate;
  final String? gender;
  final int? age;
  final bool? allowNotifications;

  User({
    required this.email,
    this.name,
    this.lastname,
    required this.role,
    required this.id,
    this.organizationId,
    this.birthdate,
    this.gender,
    this.age,
    this.allowNotifications,
  });

  factory User.fromJson(Map<String, dynamic> json) => User(
        email: json["email"],
        name: json["name"],
        lastname: json["lastname"],
        role: json["role"],
        id: json["id"],
        organizationId: json["organization_id"],
        birthdate: json["birthdate"] != null
            ? DateTime.parse(json["birthdate"])
            : null,
        gender: json["gender"],
        age: json["age"],
        allowNotifications: json["allow_notifications"],
      );

  Map<String, dynamic> toJson() => {
        "email": email,
        "name": name,
        "lastname": lastname,
        "role": role,
        "id": id,
        "organization_id": organizationId,
        "birthdate":
            "${birthdate?.year.toString().padLeft(4, '0')}-${birthdate?.month.toString().padLeft(2, '0')}-${birthdate?.day.toString().padLeft(2, '0')}",
        "gender": gender,
        "age": age,
        "allow_notifications": allowNotifications,
      };
}
